wallet-next/src/storage.rs

```diff
@@ -80,4 +86,73 @@ impl Storage {
             result[0].bytes.as_ref().unwrap().as_slice(),
         )?)
     }
+
+    pub async fn record_block(
+        &self,
+        scan_result: ScanResult,
+        nct: &mut NoteCommitmentTree,
+    ) -> anyhow::Result<Self> {
+        // Insert all new note records
+        for note_record in scan_result.new_notes {
+            let query = format!(
+                r#"INSERT INTO notes (
+                        note_commitment,
+                        height_spent,
+                        height_created,
+                        diversifier,
+                        amount,
+                        asset_id,
+                        transmission_key,
+                        blinding_factor,
+                        diversifier_index,
+                        nullifier)
+                    VALUES ({:?},NULL,{:?},{:?},{:?},{:?},{:?},{:?},{:?},{:?})"#,
+                note_record.note_commitment.0.to_bytes(),
+                //height_spent is NULL
+                scan_result.height,
+                note_record.note.diversifier().0,
+                note_record.note.amount(),
+                note_record.note.asset_id().to_bytes(),
+                note_record.note.transmission_key().0,
+                note_record.note.note_blinding().to_bytes(),
+                note_record.diversifier_index.0,
+                note_record.nullifier.to_bytes()
+            );
+
+            let _result = &self.pool.execute(&*query).await?;
```

Comment:
This is going to do each update on its own, rather than having them be done transactionally, right? Can we use a transaction instead so that other readers of the state see it ratchet forward an entire block at a time?

Reply:
I didn't see an exactly straightforward way to parameterize a set of sets of distinct values using `sqlx` here - we have to iterate over all the `NoteRecords` and map each the `NoteRecord` to the `notes` table rows, but I think we could write some dynamic SQL generation logic to build a valid query string of arbitrary length. One way we might be able to do this is to concatenate the `query` string or sub-string built for each `NoteRecord` into one long query and execute that as a single transaction.